File: android/app/src/main/java/com/neotun/android/config/XrayConfigGenerator.kt
package com.neotun.android.config

import com.neotun.android.models.*
import kotlinx.serialization.json.*

class XrayConfigGenerator {
    
    fun generateConfig(profile: VpnProfile, localPort: Int = 10808): String {
        val config = buildJsonObject {
            put("log", buildJsonObject {
                put("loglevel", "info")
            })
            
            put("inbounds", buildJsonArray {
                add(buildJsonObject {
                    put("tag", "socks-in")
                    put("protocol", "socks")
                    put("listen", "127.0.0.1")
                    put("port", localPort)
                    put("settings", buildJsonObject {
                        put("udp", true)
                    })
                })
                
                add(buildJsonObject {
                    put("tag", "http-in")
                    put("protocol", "http")
                    put("listen", "127.0.0.1")
                    put("port", localPort + 1)
                })
            })
            
            put("outbounds", buildJsonArray {
                add(generateOutbound(profile))
                
                add(buildJsonObject {
                    put("tag", "direct")
                    put("protocol", "freedom")
                })
                
                add(buildJsonObject {
                    put("tag", "blocked")
                    put("protocol", "blackhole")
                })
            })
            
            put("routing", buildJsonObject {
                put("rules", buildJsonArray {
                    add(buildJsonObject {
                        put("type", "field")
                        put("ip", buildJsonArray {
                            add("geoip:private")
                        })
                        put("outboundTag", "direct")
                    })
                })
            })
        }
        
        return Json { prettyPrint = true }.encodeToString(JsonObject.serializer(), config)
    }
    
    private fun generateOutbound(profile: VpnProfile): JsonObject {
        return buildJsonObject {
            put("tag", "proxy")
            put("protocol", profile.protocol.name.lowercase())
            
            put("settings", when (profile.protocol) {
                VpnProtocol.VMESS -> generateVMessSettings(profile)
                VpnProtocol.VLESS -> generateVLessSettings(profile)
                VpnProtocol.TROJAN -> generateTrojanSettings(profile)
                VpnProtocol.SHADOWSOCKS -> generateShadowsocksSettings(profile)
            })
            
            put("streamSettings", generateStreamSettings(profile.settings))
        }
    }
    
    private fun generateVMessSettings(profile: VpnProfile): JsonObject {
        val credentials = profile.credentials as VpnCredentials.VMess
        
        return buildJsonObject {
            put("vnext", buildJsonArray {
                add(buildJsonObject {
                    put("address", profile.server)
                    put("port", profile.port)
                    put("users", buildJsonArray {
                        add(buildJsonObject {
                            put("id", credentials.userId)
                            put("alterId", credentials.alterId)
                            put("security", credentials.security)
                        })
                    })
                })
            })
        }
    }
    
    private fun generateVLessSettings(profile: VpnProfile): JsonObject {
        val credentials = profile.credentials as VpnCredentials.VLess
        
        return buildJsonObject {
            put("vnext", buildJsonArray {
                add(buildJsonObject {
                    put("address", profile.server)
                    put("port", profile.port)
                    put("users", buildJsonArray {
                        add(buildJsonObject {
                            put("id", credentials.userId)
                            put("encryption", credentials.encryption)
                            credentials.flow?.let { put("flow", it) }
                        })
                    })
                })
            })
        }
    }
    
    private fun generateTrojanSettings(profile: VpnProfile): JsonObject {
        val credentials = profile.credentials as VpnCredentials.Trojan
        
        return buildJsonObject {
            put("servers", buildJsonArray {
                add(buildJsonObject {
                    put("address", profile.server)
                    put("port", profile.port)
                    put("password", credentials.password)
                })
            })
        }
    }
    
    private fun generateShadowsocksSettings(profile: VpnProfile): JsonObject {
        val credentials = profile.credentials as VpnCredentials.Shadowsocks
        
        return buildJsonObject {
            put("servers", buildJsonArray {
                add(buildJsonObject {
                    put("address", profile.server)
                    put("port", profile.port)
                    put("method", credentials.method)
                    put("password", credentials.password)
                })
            })
        }
    }
    
    private fun generateStreamSettings(settings: VpnSettings): JsonObject {
        return buildJsonObject {
            put("network", settings.network)
            
            if (settings.security != "none") {
                put("security", settings.security)
                
                when (settings.security) {
                    "tls" -> settings.tlsSettings?.let { 
                        put("tlsSettings", generateTlsSettings(it))
                    }
                    "reality" -> settings.realitySettings?.let {
                        put("realitySettings", generateRealitySettings(it))
                    }
                }
            }
            
            when (settings.network) {
                "ws" -> settings.wsSettings?.let {
                    put("wsSettings", generateWebSocketSettings(it))
                }
            }
        }
    }
    
    private fun generateTlsSettings(tls: TlsSettings): JsonObject {
        return buildJsonObject {
            tls.serverName?.let { put("serverName", it) }
            put("allowInsecure", tls.allowInsecure)
            if (tls.alpn.isNotEmpty()) {
                put("alpn", buildJsonArray {
                    tls.alpn.forEach { add(it) }
                })
            }
        }
    }
    
    private fun generateRealitySettings(reality: RealitySettings): JsonObject {
        return buildJsonObject {
            put("publicKey", reality.publicKey)
            put("shortId", reality.shortId)
            put("serverName", reality.serverName)
            put("fingerprint", reality.fingerprint)
        }
    }
    
    private fun generateWebSocketSettings(ws: WebSocketSettings): JsonObject {
        return buildJsonObject {
            put("path", ws.path)
            if (ws.headers.isNotEmpty()) {
                put("headers", buildJsonObject {
                    ws.headers.forEach { (key, value) ->
                        put(key, value)
                    }
                })
            }
        }
    }
}

File: windows/NeoTUN.Core/Config/XrayConfigGenerator.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using NeoTUN.Core.Models;

namespace NeoTUN.Core.Config;

public class XrayConfigGenerator
{
    public string GenerateConfig(VpnProfile profile, int localPort = 10808)
    {
        var config = new JsonObject
        {
            ["log"] = new JsonObject
            {
                ["loglevel"] = "info"
            },
            ["inbounds"] = new JsonArray
            {
                new JsonObject
                {
                    ["tag"] = "socks-in",
                    ["protocol"] = "socks",
                    ["listen"] = "127.0.0.1",
                    ["port"] = localPort,
                    ["settings"] = new JsonObject
                    {
                        ["udp"] = true
                    }
                },
                new JsonObject
                {
                    ["tag"] = "http-in",
                    ["protocol"] = "http",
                    ["listen"] = "127.0.0.1",
                    ["port"] = localPort + 1
                }
            },
            ["outbounds"] = new JsonArray
            {
                GenerateOutbound(profile),
                new JsonObject
                {
                    ["tag"] = "direct",
                    ["protocol"] = "freedom"
                },
                new JsonObject
                {
                    ["tag"] = "blocked",
                    ["protocol"] = "blackhole"
                }
            },
            ["routing"] = new JsonObject
            {
                ["rules"] = new JsonArray
                {
                    new JsonObject
                    {
                        ["type"] = "field",
                        ["ip"] = new JsonArray { "geoip:private" },
                        ["outboundTag"] = "direct"
                    }
                }
            }
        };

        return JsonSerializer.Serialize(config, new JsonSerializerOptions { WriteIndented = true });
    }

    private JsonObject GenerateOutbound(VpnProfile profile)
    {
        var outbound = new JsonObject
        {
            ["tag"] = "proxy",
            ["protocol"] = profile.Protocol.ToString().ToLowerInvariant()
        };

        outbound["settings"] = profile.Protocol switch
        {
            VpnProtocol.VMess => GenerateVMessSettings(profile),
            VpnProtocol.VLess => GenerateVLessSettings(profile),
            VpnProtocol.Trojan => GenerateTrojanSettings(profile),
            VpnProtocol.Shadowsocks => GenerateShadowsocksSettings(profile),
            _ => throw new ArgumentException($"Unsupported protocol: {profile.Protocol}")
        };

        outbound["streamSettings"] = GenerateStreamSettings(profile.Settings);

        return outbound;
    }

    private JsonObject GenerateVMessSettings(VpnProfile profile)
    {
        var credentials = (VpnCredentials.VMess)profile.Credentials;
        
        return new JsonObject
        {
            ["vnext"] = new JsonArray
            {
                new JsonObject
                {
                    ["address"] = profile.Server,
                    ["port"] = profile.Port,
                    ["users"] = new JsonArray
                    {
                        new JsonObject
                        {
                            ["id"] = credentials.UserId,
                            ["alterId"] = credentials.AlterId,
                            ["security"] = credentials.Security
                        }
                    }
                }
            }
        };
    }

    private JsonObject GenerateVLessSettings(VpnProfile profile)
    {
        var credentials = (VpnCredentials.VLess)profile.Credentials;
        
        var user = new JsonObject
        {
            ["id"] = credentials.UserId,
            ["encryption"] = credentials.Encryption
        };

        if (!string.IsNullOrEmpty(credentials.Flow))
        {
            user["flow"] = credentials.Flow;
        }

        return new JsonObject
        {
            ["vnext"] = new JsonArray
            {
                new JsonObject
                {
                    ["address"] = profile.Server,
                    ["port"] = profile.Port,
                    ["users"] = new JsonArray { user }
                }
            }
        };
    }

    private JsonObject GenerateTrojanSettings(VpnProfile profile)
    {
        var credentials = (VpnCredentials.Trojan)profile.Credentials;
        
        return new JsonObject
        {
            ["servers"] = new JsonArray
            {
                new JsonObject
                {
                    ["address"] = profile.Server,
                    ["port"] = profile.Port,
                    ["password"] = credentials.Password
                }
            }
        };
    }

    private JsonObject GenerateShadowsocksSettings(VpnProfile profile)
    {
        var credentials = (VpnCredentials.Shadowsocks)profile.Credentials;
        
        return new JsonObject
        {
            ["servers"] = new JsonArray
            {
                new JsonObject
                {
                    ["address"] = profile.Server,
                    ["port"] = profile.Port,
                    ["method"] = credentials.Method,
                    ["password"] = credentials.Password
                }
            }
        };
    }

    private JsonObject GenerateStreamSettings(VpnSettings settings)
    {
        var streamSettings = new JsonObject
        {
            ["network"] = settings.Network
        };

        if (settings.Security != "none")
        {
            streamSettings["security"] = settings.Security;

            switch (settings.Security)
            {
                case "tls" when settings.TlsSettings != null:
                    streamSettings["tlsSettings"] = GenerateTlsSettings(settings.TlsSettings);
                    break;
                case "reality" when settings.RealitySettings != null:
                    streamSettings["realitySettings"] = GenerateRealitySettings(settings.RealitySettings);
                    break;
            }
        }

        if (settings.Network == "ws" && settings.WsSettings != null)
        {
            streamSettings["wsSettings"] = GenerateWebSocketSettings(settings.WsSettings);
        }

        return streamSettings;
    }

    private JsonObject GenerateTlsSettings(TlsSettings tls)
    {
        var tlsSettings = new JsonObject
        {
            ["allowInsecure"] = tls.AllowInsecure
        };

        if (!string.IsNullOrEmpty(tls.ServerName))
        {
            tlsSettings["serverName"] = tls.ServerName;
        }

        if (tls.Alpn?.Length > 0)
        {
            var alpnArray = new JsonArray();
            foreach (var alpn in tls.Alpn)
            {
                alpnArray.Add(alpn);
            }
            tlsSettings["alpn"] = alpnArray;
        }

        return tlsSettings;
    }

    private JsonObject GenerateRealitySettings(RealitySettings reality)
    {
        return new JsonObject
        {
            ["publicKey"] = reality.PublicKey,
            ["shortId"] = reality.ShortId,
            ["serverName"] = reality.ServerName,
            ["fingerprint"] = reality.Fingerprint
        };
    }

    private JsonObject GenerateWebSocketSettings(WebSocketSettings ws)
    {
        var wsSettings = new JsonObject
        {
            ["path"] = ws.Path
        };

        if (ws.Headers?.Count > 0)
        {
            var headers = new JsonObject();
            foreach (var (key, value) in ws.Headers)
            {
                headers[key] = value;
            }
            wsSettings["headers"] = headers;
        }

        return wsSettings;
    }
}